File: AOC_2023/Day 5/5.cpp
#include <iostream>
#include <fstream>
#include <string>
#include <vector>
#include <algorithm>
#include <regex>
#include <climits>

using namespace std;

struct mapping{
    long long source, destination, range;
};

// comparator for sort mappings by source
bool cmp(mapping a, mapping b){
    return a.source < b.source;
}

class Solution
{
    public:
    vector<long long> seeds;
    vector<vector<mapping>> maps;
    void init(ifstream& input){
        string line;
        regex pattern("\\b(\\d+)\\s+(\\d+)\\s+(\\d+)\\b");
        while (getline(input, line))
        {
            if(line.find("seeds:") != string::npos){
                regex re("\\d+");
                sregex_iterator next(line.begin(), line.end(), re);
                sregex_iterator end;
                while (next != end) {
                    smatch match = *next;
                    seeds.push_back(stoll(match.str()));
                    next++;
                }
            }
            else if(line.find("map:") != string::npos){
                vector<mapping> map;
                while(getline(input, line)){
                    smatch match;
                    if(regex_search(line, match, pattern)){
                        mapping m;
                        m.destination = stoll(match[1]);
                        m.source = stoll(match[2]);
                        m.range = stoll(match[3]);
                        map.push_back(m);
                    }
                    else{
                        maps.push_back(map);
                        break;
                    }
                }
            }
        }
    }

    // just find mapping for each seed and get the location of the seed
    long long getLocation(long long seed){
        long long location = seed;
        for(auto it : maps){
            for(auto m : it){
                if(m.source <= location && location < m.source+ m.range){
                    location += m.destination - m.source;
                    break;
                }
            }
        }
        return location;
    }

    // find all possible locations for each seed and get the minimum
    void part1(){
        long long result = LLONG_MAX;
        for(auto it : seeds){
            result = min(result, getLocation(it));
        }
        cout<<"Part 1: "<< result <<"\n";
    }

    void part2(){
        long long result = LLONG_MAX;
        // sort all mappings by source
        for(auto& it : maps){
            sort(it.begin(), it.end(), cmp);
        }
        // go through each seed range (start, range length)
        for(int i = 0; i < seeds.size(); i+= 2){
            // this vector will hold all possible (start, range length) for intemediate/location for the given seed range
            vector<pair<long long, long long>> mapRanges = {{seeds[i], seeds[i+1]}};
            // go through each mapping such as seed->soil->fertilizer->water->light->temperature->humidity->location
            for(auto& it : maps){
                // this vector will hold all possible (start, range length) for current intemediate(mapping) for the seed
                vector<pair<long long, long long>> newRanges;
                // go through each value in mapRanges and update it based on current mapping
                for(auto& t : mapRanges){
                    // this vector will hold all possible ranges for current mapping for the seed if there's any intersection or not
                    vector<pair<long long, long long>> temp;
                    // go through each mapping and update the (start, range length)
                    for(auto& m : it){
                        long long offset = m.destination - m.source;
                        // if there's overlap between current mapping and current (start, range length)
                        if(t.first >= m.source && t.first < m.source + m.range){
                            // if the current mapping covers the whole range
                            if(m.source + m.range >= t.first + t.second){
                                temp.push_back({t.first + offset, t.second});
                            }
                            // if the current mapping covers only part of the (start, range length)
                            else{
                                // update the range by updating start and range length and add the overlapped part to temp
                                t.second -= (m.source + m.range - t.first);
                                temp.push_back({t.first + offset, m.source + m.range - t.first});
                                t.first = m.source + m.range;
                            }
                        }
                    }
                    // if there's no intersection between current mapping and current range, just add the current range to temp
                    if(temp.empty()){
                        newRanges.push_back(t);
                    }
                    // if there's intersection, add all possible ranges to newRanges
                    else{
                        newRanges.insert(newRanges.end(), temp.begin(), temp.end());
                    }
                }
                // update mapRanges with newRanges
                mapRanges = newRanges;
            }
            // get the minimum location for the current seed range
            auto minElement = std::min_element(mapRanges.begin(), mapRanges.end(),
            [](const auto& a, const auto& b) {
                return a.first < b.first;
            }); 
            // update the result
            result = min(minElement->first, result);
        }
        cout<<"Part 2: "<< result <<"\n";
    }

};

int main(){
    ifstream in("input.txt");
    if (!in.is_open())
    {
        cerr << "Could not open the file - '" << endl;
        return EXIT_FAILURE;
    }
    Solution s;
    s.init(in);
    s.part1();
    s.part2();
    return 0;
}

File: AOC_2023/Day 5/5.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"regexp"
	"strings"
	"strconv"
	"math"
	"sort"
)

type mapping struct {
	dest int64
	src int64
	length int64
}

func Less(a, b mapping) bool {
	return a.src < b.src
}

func initialize(filename string) ([]int64, [][]mapping, error) {
	file, err := os.Open(filename)
	if err != nil {
		return nil, nil, err
	}
	defer file.Close()
	reader := bufio.NewReader(file)
	pattern := regexp.MustCompile("\\b(\\d+)\\s+(\\d+)\\s+(\\d+)\\b")
	var seeds []int64
	var maps [][]mapping
	for {
		line, err := reader.ReadString('\n')
		if err != nil {
			break
		}
		if strings.Contains(line, "seeds:") {
			re := regexp.MustCompile("\\d+")
			for _, match := range re.FindAllString(line, -1) {
				num, _ := strconv.ParseInt(match, 10, 64)
				seeds = append(seeds, num)
			}
		}
		if strings.Contains(line, "map:") {
			line, _ = reader.ReadString('\n')
			temp := []mapping{}
			for pattern.MatchString(line) {
				matches := pattern.FindStringSubmatch(line)
				dest, _ := strconv.ParseInt(matches[1], 10, 64)
				src, _ := strconv.ParseInt(matches[2], 10, 64)
				length, _ := strconv.ParseInt(matches[3], 10, 64)
				temp = append(temp, mapping{dest, src, length})
				line, _ = reader.ReadString('\n')
			}
			maps = append(maps, temp)
		}
	}
	return seeds, maps, nil
}

func part1(seeds []int64, maps [][]mapping) {
	var result int64 = math.MaxInt64
	for _, seed := range seeds {
		temp := seed
		for _, m := range maps {
			for _, mapping := range m {
				if mapping.src <= temp && temp < mapping.src + mapping.length {
					temp += mapping.dest - mapping.src
					break
				}
			}
		}
		result = min(result, temp)
	}
	fmt.Println("Part 1:", result)
}

func part2(seeds []int64, maps [][]mapping) {
	var result int64 = math.MaxInt64
	for _, m := range maps {
		sort.Slice(m, func(i, j int) bool {
			return Less(m[i], m[j])
		})
	}
	//fmt.Println(maps)
	for i := 0; i < len(seeds); i+=2 {
		mapRanges := map[int64]int64{ seeds[i]: seeds[i+1] }
		for _, m := range maps {
			newRanges := map[int64]int64{}
			for src, length := range mapRanges {
				temp := map[int64]int64{}
				for _, mapping := range m {
					offset := mapping.dest - mapping.src
					if mapping.src <= src && src < mapping.src + mapping.length {
						if mapping.src + mapping.length >= src + length {
							temp[src + offset] = length
						} else {
							length -= (mapping.src + mapping.length - src)
							temp[src + offset] = mapping.src + mapping.length - src
							src = mapping.src + mapping.length
						}
					}
				}
				if len(temp) == 0 {	
					newRanges[src] = length
				} else {
					for key, value := range temp {
						newRanges[key] = value
					}
				}
			}
			mapRanges = map[int64]int64{}
			for src, length := range newRanges {
				mapRanges[src] = length
			}
		}
		var minSource int64 = math.MaxInt64
		for src := range mapRanges {
			minSource = min(minSource, src)
		}
		result = min(result, minSource)
	}
	fmt.Println("Part 2:", result)
}

func main() {
	filename := "input.txt"
	seeds, maps, err := initialize(filename)
	if err != nil {
		fmt.Println("Error reading file:", err)
		return
	}
	part1(seeds, maps)
	part2(seeds, maps)
}